File: concepts/golang/queue.go
package main

import "fmt"

// Queue represents a basic queue data structure.
type Queue struct {
	items []interface{}
}

// Enqueue adds an item to the end of the queue.
func (q *Queue) Enqueue(item interface{}) {
	q.items = append(q.items, item)
}

// Dequeue removes and returns the front item from the queue.
func (q *Queue) Dequeue() interface{} {
	if len(q.items) == 0 {
		return nil
	}
	frontItem := q.items[0]
	q.items = q.items[1:]
	return frontItem
}

// Front returns the front item from the queue without removing it.
func (q *Queue) Front() interface{} {
	if len(q.items) == 0 {
		return nil
	}
	return q.items[0]
}

// IsEmpty returns true if the queue is empty.
func (q *Queue) IsEmpty() bool {
	return len(q.items) == 0
}

// Size returns the number of items in the queue.
func (q *Queue) Size() int {
	return len(q.items)
}

func main() {
	queue := Queue{}

	queue.Enqueue(10)
	queue.Enqueue(20)
	queue.Enqueue(30)

	fmt.Println("Size:", queue.Size())
	fmt.Println("Front:", queue.Front())

	fmt.Println("Dequeue:", queue.Dequeue())
	fmt.Println("Size:", queue.Size())

	fmt.Println("Dequeue:", queue.Dequeue())
	fmt.Println("IsEmpty:", queue.IsEmpty())

	fmt.Println("Dequeue:", queue.Dequeue())
	fmt.Println("IsEmpty:", queue.IsEmpty())
}

File: practiceProblems/dsa/queue.py
class Queue:
    '''
    A simple queue implementation using a list.
    Supports basic operations: enqueue, dequeue, peek, is_empty, size, and display.
    TC: O(1) for enqueue and O(n) for dequeue
    SC: O(n) for storing n elements
    '''
    def __init__(self):
        self.queue = []

    def enqueue(self, item):
        self.queue.append(item)
        print(f"Enqueued {item}.")

    def dequeue(self):
        if self.is_empty():
            print("Queue is empty. Cannot dequeue.")
            return None
        return self.queue.pop(0)  # O(n) operation

    def peek(self):
        if self.is_empty():
            print("Queue is empty. Nothing to peek.")
            return None
        return self.queue[0]

    def is_empty(self):
        return len(self.queue) == 0

    def size(self):
        return len(self.queue)

    def display(self):
        print("Queue (front -> rear):", self.queue)

q = Queue()
q.enqueue(5)
q.enqueue(15)
q.enqueue(25)
q.display()
print("Front item:", q.peek())
print("Dequeued item:", q.dequeue())
q.display()
